### DISCUSSION TITLE
OTLP | run grpc_metrics_main.cc

### DISCUSSION THREAD

**Discussion by yonatan555** [open-telemetry/opentelemetry-cpp#1776] [Q&A]
Hi I built a otlp project with grpc, metrics, prometheues with the this command :
cmake -DWITH_EXAMPLES=OFF -DBUILD_TESTING=OFF -DWITH_OTLP=ON -DWITH_PROMETHEUS=ON -DWITH_OTLP_GRPC=ON -DCMAKE_PERFIX_PATH=/root/grpc/ .

and got errors about the linker.

part of them:

`
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: ./lib/libopentelemetry_proto.a(metrics_service.grpc.pb.cc.o): in function `grpc::ClientAsyncResponseReader<opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceResponse>* grpc::internal::ClientAsyncResponseReaderHelper::Create<opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceResponse, opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceRequest, google::protobuf::MessageLite, google::protobuf::MessageLite>(grpc::ChannelInterface*, grpc::CompletionQueue*, grpc::internal::RpcMethod const&, grpc::ClientContext*, opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceRequest const&)':
metrics_service.grpc.pb.cc:(.text._ZN4grpc8internal31ClientAsyncResponseReaderHelper6CreateIN13opentelemetry5proto9collector7metrics2v128ExportMetricsServiceResponseENS7_27ExportMetricsServiceRequestEN6google8protobuf11MessageLiteESC_EEPNS_25ClientAsyncResponseReaderIT_EEPNS_16ChannelInterfaceEPNS_15CompletionQueueERKNS0_9RpcMethodEPNS_13ClientContextERKT0_[_ZN4grpc8internal31ClientAsyncResponseReaderHelper6CreateIN13opentelemetry5proto9collector7metrics2v128ExportMetricsServiceResponseENS7_27ExportMetricsServiceRequestEN6google8protobuf11MessageLiteESC_EEPNS_25ClientAsyncResponseReaderIT_EEPNS_16ChannelInterfaceEPNS_15CompletionQueueERKNS0_9RpcMethodEPNS_13ClientContextERKT0_]+0xab): undefined reference to `grpc::g_core_codegen_interface'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: metrics_service.grpc.pb.cc:(.text._ZN4grpc8internal31ClientAsyncResponseReaderHelper6CreateIN13opentelemetry5proto9collector7metrics2v128ExportMetricsServiceResponseENS7_27ExportMetricsServiceRequestEN6google8protobuf11MessageLiteESC_EEPNS_25ClientAsyncResponseReaderIT_EEPNS_16ChannelInterfaceEPNS_15CompletionQueueERKNS0_9RpcMethodEPNS_13ClientContextERKT0_[_ZN4grpc8internal31ClientAsyncResponseReaderHelper6CreateIN13opentelemetry5proto9collector7metrics2v128ExportMetricsServiceResponseENS7_27ExportMetricsServiceRequestEN6google8protobuf11MessageLiteESC_EEPNS_25ClientAsyncResponseReaderIT_EEPNS_16ChannelInterfaceEPNS_15CompletionQueueERKNS0_9RpcMethodEPNS_13ClientContextERKT0_]+0xc1): undefined reference to `grpc::g_core_codegen_interface'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: ./lib/libopentelemetry_proto.a(metrics_service.grpc.pb.cc.o): in function `grpc::ClientAsyncResponseReader<opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceResponse>::StartCall()':
metrics_service.grpc.pb.cc:(.text._ZN4grpc25ClientAsyncResponseReaderIN13opentelemetry5proto9collector7metrics2v128ExportMetricsServiceResponseEE9StartCallEv[_ZN4grpc25ClientAsyncResponseReaderIN13opentelemetry5proto9collector7metrics2v128ExportMetricsServiceResponseEE9StartCallEv]+0x1f): undefined reference to `grpc::g_core_codegen_interface'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: metrics_service.grpc.pb.cc:(.text._ZN4grpc25ClientAsyncResponseReaderIN13opentelemetry5proto9collector7metrics2v128ExportMetricsServiceResponseEE9StartCallEv[_ZN4grpc25ClientAsyncResponseReaderIN13opentelemetry5proto9collector7metrics2v128ExportMetricsServiceResponseEE9StartCallEv]+0x33): undefined reference to `grpc::g_core_codegen_interface'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: ./lib/libopentelemetry_proto.a(metrics_service.grpc.pb.cc.o): in function `grpc::internal::RpcMethodHandler<opentelemetry::proto::collector::metrics::v1::MetricsService::Service, opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceRequest, opentelemetry::proto::collector::metrics::v1::ExportMetricsServiceResponse, google::protobuf::MessageLite, google::protobuf::MessageLite>::Deserialize(grpc_call*, grpc_byte_buffer*, grpc::Status*, void**)':
metrics_service.grpc.pb.cc:(.text._ZN4grpc8internal16RpcMethodHandlerIN13opentelemetry5proto9collector7metrics2v114MetricsService7ServiceENS6_27ExportMetricsServiceRequestENS6_28ExportMetricsServiceResponseEN6google8protobuf11MessageLiteESD_E11DeserializeEP9grpc_callP16grpc_byte_bufferPNS_6StatusEPPv[_ZN4grpc8internal16RpcMethodHandlerIN13opentelemetry5proto9collector7metrics2v114MetricsService7ServiceENS6_27ExportMetricsServiceRequestENS6_28ExportMetricsServiceResponseEN6google8protobuf11MessageLiteESD_E11DeserializeEP9grpc_callP16grpc_byte_bufferPNS_6StatusEPPv]+0x24): undefined reference to `grpc::g_core_codegen_interface'
/opt/rh/devtoolset-9/root/usr/libexec/gcc/x86_64-redhat-linux/9/ld: ./lib/libopentelemetry_proto.a(metrics_service.grpc.pb.cc.o):metrics_service.grpc.pb.cc:(.text._ZN4grpc8internal16RpcMethodHandlerIN13opentelemetry5proto9collector7metrics2v114MetricsService7ServiceENS6_27ExportMetricsServiceRequestENS6_28ExportMetricsServiceResponseEN6google8protobuf11MessageLiteESD_E11DeserializeEP9grpc_callP16grpc_byte_bufferPNS_6StatusEPPv[_ZN4grpc8internal16RpcMethodHandlerIN13opentelemetry5proto9collector7metrics2v114MetricsService7ServiceENS6_27ExportMetricsServiceRequestENS6_28ExportMetricsServiceResponseEN6google8protobuf11MessageLiteESD_E11DeserializeEP9grpc_callP16grpc_byte_bufferPNS_6StatusEPPv]+0x3a): more undefined references to `grpc::g_core_codegen_interface' follow

`

**Comment by marcalff** [comment]
```-DCMAKE_PERFIX_PATH=/root/grpc/```

Not sure if this is the problem, but in any case, it should be `CMAKE_PREFIX_PATH`.

**Reply by yonatan555** [reply]
Yeah that was part of the problem, and I understand that there are libs which missing, I am trying to find which exactly.

**Reply by yonatan555** [reply]
Hi, i did it and the a lot of static libs, grpc++, grpcpp,grpc and etc...
How can I know that I bulit the all libs that are needed, for otlp, 
Because I still have the bugs above.

**Reply by owent** [reply]
You can find paths of all depended packages in `<Build dir>/CMakeCache.txt` after run `cmake <Build dir> ...`, and check if all the packages are correct.

**Reply by yonatan555** [reply]
The packages are installed automatically by the gGRP cmake?
Thanks

**Reply by owent** [reply]
No, you should build or install it yourself.
